### ISSUE TITLE
"GET %2Ffavicon.ico HTTP/1.1" in log

### ISSUE THREAD

**Issue by znz** [rails/rails#11816]
How to reproduce:

1. `rails new foo` (I use rails 4.0.0)
2. edit `Gemfile` and enable `unicorn` gem
3. run `bundle exec unicorn`
4. open `http://localhost:8080/` with browser.

then I found `"GET %2Ffavicon.ico HTTP/1.1"` in log messages.
I expected `"GET /favicon.ico HTTP/1.1"`.

In `actionpack-4.0.0/lib/action_dispatch/middleware/static.rb`,
`match = @file_handler.match?(path)` returns escaped path,
and overwrite `env["PATH_INFO"]`.
It seems to affect logger.

**Comment by tenderlove** [comment]
@jeremy the ruby-lang bug you linked to in rack/rack#265 is broken. Are you able to find the right URL?  Since Rack master is 2.0, I'd like to fix this The Right Way™

**Comment by matthewd** [comment]
@tenderlove see http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/24014 and surrounds.

AFAICS, it's a bug report complaining that `URI.encode` works the way @jeremy is wants (`+` means `+`), while the reporter expected it to work the way Rack currently does (`+` means ` `), and @shyouhei pointing out that the current behaviour matches the RFC.

That matches my understanding too:

> **Within the query string**, the plus sign is reserved as shorthand notation for a space. Therefore, real plus signs must be encoded.

(emphasis mine, but the heading's hardly subtle in [the original](http://www.w3.org/Addressing/URL/4_URI_Recommentations.html#z5))

[RFC 3986](https://tools.ietf.org/html/rfc3986) ascribes no special interpretation to `+`. In fact, it carries an [example](https://tools.ietf.org/html/rfc3986#section-1.1.2) to the contrary: in `tel:+1-816-555-1212`, `+` clearly means `+`.

**Comment by tenderlove** [comment]
@matthewd So we should really only be dealing with `+` encoding in the query string, and not the path.

Shouldn't we just change Rack to use URI unescape vs form unescape?  It looks like [Rack is using form unescape rather than URI escape](https://github.com/rack/rack/blob/9792bf590add0f09f301f00cdb296b333ec01f24/lib/rack/utils.rb#L44-L46):

```
irb(main):008:0> Rack::Utils.unescape("/foo+bar")
=> "/foo bar"
irb(main):009:0> URI.decode_www_form_component("/foo+bar")
=> "/foo bar"
irb(main):010:0> URI::Parser.new.unescape "/foo+bar"
=> "/foo+bar"
irb(main):011:0> 
```

As a side note, [this comment seems dubious / wrong](https://github.com/rack/rack/blob/9792bf590add0f09f301f00cdb296b333ec01f24/lib/rack/utils.rb#L35-L40).

It seems like we have an `escape` and an `unescape` (for query parameters and form components) and an `escape_path` for path components, but no `unescape_path`.  Does this patch seem reasonable?

```diff
diff --git a/lib/rack/file.rb b/lib/rack/file.rb
index 1efa15a..5b755f5 100644
--- a/lib/rack/file.rb
+++ b/lib/rack/file.rb
@@ -30,7 +30,7 @@ module Rack
         return fail(405, "Method Not Allowed", {'Allow' => ALLOW_HEADER})
       end
 
-      path_info = Utils.unescape request.path_info
+      path_info = Utils.unescape_path request.path_info
       clean_path_info = Utils.clean_path_info(path_info)
 
       path = ::File.join(@root, clean_path_info)
diff --git a/lib/rack/utils.rb b/lib/rack/utils.rb
index 184bf5a..8911445 100644
--- a/lib/rack/utils.rb
+++ b/lib/rack/utils.rb
@@ -19,6 +19,8 @@ module Rack
     COMMON_SEP = QueryParser::COMMON_SEP
     KeySpaceConstrainedParams = QueryParser::Params
 
+    URI_PARSER = URI::Parser.new
+
     class << self
       attr_accessor :default_query_parser
     end
@@ -35,10 +37,17 @@ module Rack
     # Like URI escaping, but with %20 instead of +. Strictly speaking this is
     # true URI escaping.
     def escape_path(s)
-      escape(s).gsub('+', '%20')
+      URI_PARSER.escape s
     end
     module_function :escape_path
 
+    # Unescapes the **path** component of a URI.  See Rack::Utils.unescape for
+    # unescaping query parameters or form components.
+    def unescape_path(s)
+      URI_PARSER.unescape s
+    end
+    module_function :unescape_path
+
     # Unescapes a URI escaped string with +encoding+. +encoding+ will be the
     # target encoding of the string returned, and it defaults to UTF-8
     def unescape(s, encoding = Encoding::UTF_8)
```

/cc @jeremy

**Comment by tenderlove** [comment]
Fixed in e25fdad2f147e6f368958f9a06a5ac9d10288408
